# LaboratoryTask2 

**Jade Natasha S. Ragot**

**DS4A**



**Laboratory Task 2**

*(01_Understanding Deep Learning)*


<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a single forward pass and compute for the error.
</p>

$$x = \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix}$$

$$y = \begin{bmatrix} 1 \end{bmatrix}$$

$$f=max(0, Z_n)$$

$$\text{hidden unit weights} =
\begin{bmatrix}
w_{11} = 0.2 & w_{12} = -0.3 \\
w_{13} = 0.4 & w_{14} = 0.1 \\
w_{15} = -0.5 & w_{16} = 0.2
\end{bmatrix}$$

$$\text{output unit weights} =
\begin{bmatrix}
w_{21} = -0.3 \\
w_{22} = -0.2
\end{bmatrix}$$

$$\theta =
\begin{bmatrix}
\theta_{1} = -0.4 \\
\theta_{2} = 0.2 \\
\theta_{3} = 0.1
\end{bmatrix}$$

**Given**

Inputs: $x_1$= 1, $x_2$ = 0, $x_3$ = 1

Target Output: y = 1

Hidden Weights: $w_{11}$ = 0.2, $w_{12}$ = -0.3, $w_{13}$ = 0.4, $w_{14}$ = 0.1, $w_{15}$ = -0.5, $w_{16}$ = 0.2

Output Weights: $w_{21}$ = -0.3, $w_{22}$ = -0.2

Biases: $ \theta_1$ = -0.4, $ \theta_2$ = 0.2, $ \theta_3$ = 0.1

*Weighted sum for Z1 (uses w11, w13, w15 — the weights feeding into hidden neuron 1)*

>>  Z1 = (w11 * x1) + (w13 * x2) + (w15 * x3) + theta1

*Weighted sum for Z2 (uses w12, w14, w16 — the weights feeding into hidden neuron 2)*

>> Z2 = (w12 * x1) + (w14 * x2) + (w16 * x3) + theta2

In [12]:
import numpy as np

# Computing the weighted sums into the hidden layer (Z1, Z2)

x = np.array([1, 0, 1])          # inputs
W = np.array([
    [0.2, -0.3],    # weights from x1 -> [H1, H2]
    [0.4,  0.1],    # weights from x2 -> [H1, H2]
    [-0.5, 0.2]     # weights from x3 -> [H1, H2]
])
theta = np.array([-0.4, 0.2])    # biases for [H1, H2]

Z = x @ W + theta   # matrix multiply + add bias

In [11]:
# Applying the activation function (ReLU) to get H1, H2
# ReLU (Rectified Linear Unit) rule: negative → 0, positive → keep it.

def relu(z):
    return np.maximum(0, z)

H = relu(Z)     # not relu(Z_hidden)
print("H1, H2 =", H)

H1, H2 = [0.  0.1]


We can observe that H1 got zeroed out because its input sum was negative.

In [13]:
# Computing the weighted sum into the output layer (Z3)
# Now H1 and H2 are the inputs, and w21, w22 are the connecting weights.

w21, w22 = -0.3, -0.2
theta3 = 0.1

Z3 = (w21 * H[0]) + (w22 * H[1]) + theta3
print("Z3 =", Z3)

Z3 = 0.08


In [14]:
# Applying activation to get the predicted output ŷ

y_hat = relu(Z3)
print("y_hat =", y_hat)

y_hat = 0.08


So the network predicts ŷ = 0.08, while the actual target is y = 1.

In [19]:
# Computing the error
y_actual = 1

E = 0.5 * (y_actual - y_hat) ** 2
print("So the Error is E =", E)

So the Error is E = 0.4232


The whole point of computing this error — E = 0.4232 — is to measure "how wrong is the network right now?" This large error tells us that the weights need to be adjusted to get closer to the target.

Since we are still in the feedforward (forward pass) stage, the model has not yet been trained — it hasn't learned anything yet. The weights and biases used here were just the initial/given values, not values the network arrived at through learning.